### PR TITLE
batch chunk data

### PR DESCRIPTION
**Pull request by zhexuany**
This PR brings an optimization which can improve performance by 20% against tikv. 

`TiChunkBatchColumnVector` is added to batch chunk data. 

We also add a conf `CHUNK_BATCH_SIZE` which allows user can adjust the batch size as they need.

### CODE REVIEW

File: core/src/main/scala/com/pingcap/tispark/TiConfigConst.scala [hunk 1]
```diff
@@ -30,6 +30,7 @@ object TiConfigConst {
   val COPROCESS_STREAMING: String = "spark.tispark.coprocess.streaming"
   val ENABLE_CHUNK: String = "spark.tispark.coprocess.chunk"
   val UNSUPPORTED_PUSHDOWN_EXPR: String = "spark.tispark.plan.unsupported_pushdown_exprs"
+  val CHUNK_BATCH_SIZE: String = "spark.tispark.coprocessor.chunk_batch_size"
```
Comment (marsishandsome) [inline core/src/main/scala/com/pingcap/tispark/TiConfigConst.scala]: add this config to `README.md`?

File: tikv-client/src/main/java/com/pingcap/tikv/columnar/TiChunkBatchColumnVector.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+ * limitations under the License.
+ */
+package com.pingcap.tikv.columnar;
+
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.List;
+
+/** An implementation of {@link TiColumnVector}. All data is stored in TiDB chunk format. */
+public class TiChunkBatchColumnVector extends TiColumnVector {
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/columnar/TiChunkBatchColumnVector.java]: How about BatchedTiColumnVector?

File: tikv-client/src/main/java/com/pingcap/tikv/operation/iterator/CoprocessorIterator.java [hunk 3]
```diff
@@ -132,13 +134,26 @@ public TiChunk next() {
           return new TiChunk(columnarVectors);
         } else {
           // hasNext => create dataInput, so we do not need to advance next dataInput.
-          TiChunkColumnVector[] columnarVectors = new TiChunkColumnVector[dataTypes.length];
+          TiColumnVector[] columnarVectors = new TiColumnVector[dataTypes.length];
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/operation/iterator/CoprocessorIterator.java]: It is possible if user sets a very large batch size.

File: core/src/main/scala/com/pingcap/tispark/utils/TiUtil.scala [hunk 4]
```diff
@@ -209,6 +209,9 @@ object TiUtil {
     tiConf
   }
 
+  def getChunkBatchSize(sqlContext: SQLContext) =
+    sqlContext.getConf(TiConfigConst.CHUNK_BATCH_SIZE, "1024").toInt
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply core/src/main/scala/com/pingcap/tispark/utils/TiUtil.scala]: In `TiDBRelation`, there is no `SparkConf` available.

File: tikv-client/src/main/java/com/pingcap/tikv/columnar/TiChunkBatchColumnVector.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (birdstorm) [reply tikv-client/src/main/java/com/pingcap/tikv/columnar/TiChunkBatchColumnVector.java]: sure

File: tikv-client/src/main/java/com/pingcap/tikv/operation/iterator/CoprocessorIterator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (birdstorm) [reply tikv-client/src/main/java/com/pingcap/tikv/operation/iterator/CoprocessorIterator.java]: let's try to add a limit for the estimated size(length * sum_of_all_data_types)

File: tikv-client/src/main/java/com/pingcap/tikv/columnar/TiChunkBatchColumnVector.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+  private int[] getColumnVectorIdxAndRowId(int rowId) {
+    int offset = Arrays.binarySearch(this.rightEndpoints, rowId);
+    int idx;
+    if (offset >= 0) {
+      idx = offset;
+    } else {
+      idx = -(offset + 2);
+    }
+    assert idx < childColumns.size() && idx >= 0;
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/columnar/TiChunkBatchColumnVector.java]: DONE

File: core/src/main/scala/com/pingcap/tispark/TiConfigConst.scala (hunk 1 above)
Comment (zhexuany) [reply core/src/main/scala/com/pingcap/tispark/TiConfigConst.scala]: DONE

**Comment by marsishandsome** [comment]
```
This PR brings an optimization which can improve performance by 20% against tikv.
```
do you have benchmark result in detail?

File: README.md [hunk 6]
```diff
@@ -205,6 +205,7 @@ The configurations in the table below can be put together with `spark-defaults.c
 | `spark.tispark.show_rowid` |  `false` | Whether to show the implicit row ID if the ID exists |
 | `spark.tispark.db_prefix` |  `""` | The string that indicates the extra prefix for all databases in TiDB. This string distinguishes the databases in TiDB from the Hive databases with the same name. |
 | `spark.tispark.request.isolation.level` |  `SI` | Isolation level means whether to resolve locks for the underlying TiDB clusters. When you use the "RC", you get the latest version of record smaller than your `tso` and ignore the locks. If you use "SI", you resolve the locks and get the records depending on whether the resolved lock is committed or aborted.  |
+| `spark.tispark.coprocessor.chunk_batch_size` | `1024` | How many rows fetched from Coprocessor |
```
Comment (marsishandsome) [inline README.md]: would u pls tell me how you choose the default value `1024`?

Comment (zhexuany) [reply README.md]: The size of L3 cache is 4096, so 1024, 2048, 4096 are reasonable numbers.

**Comment by zhexuany** [comment]
> ```
> This PR brings an optimization which can improve performance by 20% against tikv.
> ```
> 
> 
> do you have benchmark result in detail?

I already shared the benchmark to all team members.

**Comment by zhexuany** [comment]
> ```
> This PR brings an optimization which can improve performance by 20% against tikv.
> ```
> 
> 
> do you have benchmark result in detail?

I already shared the benchmark to all team members.

**Pull request merged**